portal/views/auth.py

```diff
@@ -69,6 +79,38 @@
 )
 
 
+class KeyForm(FlaskForm):
+    key = IntegerField('key', validators=[DataRequired()])
+
+    def validate_key(form, field):
+        """Use form validation to verify token"""
+        user = unchecked_get_user(session.get('user_needing_2fa'))
+
+        if user.validate_otp(field.data):
+            current_app.logger.debug(f"{user.id} passed 2FA with valid token")
+        else:
+            current_app.logger.debug(f"{user.id} failed 2FA token validation")
+            raise ValidationError("invalid token")
+
+
+@auth.route('/2fa/verify', methods=('GET', 'POST'))
+def two_factor_auth():
+    try:
+        user = unchecked_get_user(session.get('user_needing_2fa'))
+    except (BadRequest, ValueError):
+        abort(404)
+
+    form = KeyForm()
+    if not form.validate_on_submit():
+        return render_template('flask_user/second_factor.html', form=form)
+
+    # Still here implies validation was successful, resume interrupted login
+    session.pop('user_needing_2fa')
+    session['2FA_verified'] = '2FA verified'
```

Comment:
perhaps as an extra layer of paranoia, this should be the user_id that passed 2FA?